packages/gator-permissions-snap/src/core/permissionRequestLifecycleOrchestrator.ts

```diff
@@ -165,10 +164,26 @@ export class PermissionRequestLifecycleOrchestrator {
 
       await confirmationDialog.updateContent({
         ui,
-        isGrantDisabled: isGrantDisabled || hasValidationErrors,
+        isGrantDisabled: isGrantDisabled || hasValidationErrors(metadata),
       });
     };
 
+    // Set up validation callback that runs when grant button is clicked.
+    // Race condition scenario this prevents:
+    //   1. User types invalid input → validation debounced (500ms delay)
+    //   2. User clicks Grant before 500ms elapses (button still enabled)
+    //   3. Button click event flushes all pending debounced events
+    //   4. Validation runs → updates UI with errors & disables button
+    //   5. Button handler already invoked (button was enabled at click time)
+    //   6. This callback catches it → returns false → dialog stays open
+    confirmationDialog.setBeforeGrantCallback(async () => {
+      const metadata = await lifecycleHandlers.deriveMetadata({ context });
+      return !hasValidationErrors(metadata);
+    });
```

Comment:
instead of this setBeforeGrantCallback, could we make the `beforeGrantCallback` a required constructor argument? The `let context` variable is always updated to the latest state, so afaict everything required for this callback should be referencable at the time that the confirmation is constructed.